webrtc/README.md

```diff
@@ -0,0 +1,416 @@
+[multihash](https://github.com/multiformats/multihash).
+
+For compatibility implementations MUST support hash algorithm
+[`sha-256`](https://github.com/multiformats/multihash) and base encoding
+[`base64url`](https://github.com/multiformats/multibase). Implementations MAY
+support other hash algorithms and base encodings, but they may not be able to
+connect to all other nodes.
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+1. Server node _B_ generates a TLS certificate, listens on a UDP port and
+   advertises the corresponding multiaddress (see [#Addressing]) through some
+   external mechanism.
+
+   Given that _B_ is publicly reachable, _B_ acts as a [ICE
+   Lite](https://www.rfc-editor.org/rfc/rfc5245) agent. It listens on a UDP port
+   for incoming STUN and SCTP packets and multiplexes based on source IP and
+   source port.
+
+2. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+3. _A_ instantiates a `RTCPeerConnection`. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+   _A_ MAY reuse the same certificate across `RTCPeerConnection`s. Though one
+   should note that reusing the certificate can be used to identify _A_ across
+   connections by on-path observers given that WebRTC uses TLS 1.2. Thus this
+   specification RECOMMENDS that _A_, i.e. the browser, does not reuse the same
+   certificate across `RTCPeerConnection`.
+
+4. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+5. _A_ creates a local offer via
+   [`RTCPeerConnection.createOffer()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createOffer).
+   _A_ generates a random string and sets that string as the username (_ufrag_
+   or _username fragment_) and password on the SDP of the local offer. Finally
+   _A_ sets the modified offer via
+   [`RTCPeerConnection.setLocalDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setLocalDescription).
+
+   Note that this process, oftentimes referred to as "SDP munging" is disallowed
+   by the specification, but not enforced across the major browsers (Safari,
+   Firefox, Chrome) due to use-cases in the wild. See also
+   https://bugs.chromium.org/p/chromium/issues/detail?id=823036
+
+6. _A_ establishes the connection to _B_. The random string used as a _username_
+   and _password_ in combination with the IP and port of _A_ can be used by _B_
```

Comment:
At what point is the random string that is used as the _username_ and _password_ sent by **A** to **B**?